### PR TITLE
Add with_sharding_constraint method to be used within sharded_jit.

### PR DESCRIPTION
**Pull request by skye**
See the with_sharding_constraint docstring for a description of what this method does.

Depending on how we decide nested sharded_jits should work, an
alternative implementation for with_sharding_constraint could be:
```python
def with_sharding_constraint(x, partitions):
    return sharded_jit(lambda x: x, in_parts=partitions, out_parts=partitions)
```
In this case, we could get rid of the with_sharding_constraint primitive, and
possibly even the API. This implementation gets the job done for now
without committing to a nested sharded_jit behavior, and is also much
easier to take the gradient of than sharded_jit.

### CODE REVIEW

**Comment by skye** [comment]
This includes https://github.com/google/jax/pull/3099. Look at the diff for  https://github.com/google/jax/pull/3100/commits/fd65f0fd0b45d374f55fca986029c465de2ad81a for this change only (or wait until I merge #3099).

File: jax/interpreters/sharded_jit.py
Comment (jekbradbury) [inline jax/interpreters/sharded_jit.py:220]:
```suggestion
  """Returns the total number of partitions from PartitionSpecs inside `jaxpr`.
```

File: jax/interpreters/sharded_jit.py
Comment (jekbradbury) [inline jax/interpreters/sharded_jit.py:231]: That definitely seems like better UX, but I'm also wary of bringing too much about pmap/sharding/etc into the JaxprTrace machinery. For maps, what I was thinking (very tentatively) was to create a final-style, dynamically-scoped "SPMD tracer" that maintains the axis context, propagates replication, handles axis_index, etc.; if you want the current pmap behavior, you can trace the output of that to an SPMD jaxpr (and translate that to SPMD XLA), but you can also wrap it in axis-consuming traces for other kinds of maps (vmap, soft_pmap, op-by-op pmap). Maybe something like that makes sense for partitions; maybe not. I guess all I'm saying is: I don't think we should necessarily worry about fixing this part now/soon 🙂 .

File: jax/interpreters/sharded_jit.py
Comment (jekbradbury) [inline jax/interpreters/sharded_jit.py:385]: Is there a way to make it clear that this isn't a "please reshard this value from however it was sharded before to this new sharding" (even if maybe it could eventually mean that in a hypothetical op-by-op context) but more like "annotate this value as a constraint for the compiler, so that it's sharded in this way at runtime and the compiler will propagate that sharding forwards and/or backwards as it sees fit"?

**Review by jekbradbury — APPROVED** [review]

File: jax/interpreters/sharded_jit.py
Comment (skye) [reply jax/interpreters/sharded_jit.py:231]: Agreed we shouldn't bake any of this logic into the core tracing machinery. I like your tentative SPMD tracer idea. I was vaguely imagining using dynamic-scoped tracing and/or expanding pmap's DynamicAxisEnv concept to do final-style tracing for all parallel constructs, but it might be too complicated. I think we'll get more clarity as we expand the current initial-style transforms to compose with each other.

File: jax/interpreters/sharded_jit.py
Comment (skye) [reply jax/interpreters/sharded_jit.py:393]: I added a bunch more verbiage in an attempt to explain this :) Lemme know if you have any suggestions to make it more clear.

File: jax/interpreters/sharded_jit.py
Comment (hawkinsp) [inline jax/interpreters/sharded_jit.py:356]: `with_sharding_constraint` ?

Might be a better name since it gives the intuition that information can flow backwards.

File: jax/interpreters/sharded_jit.py
Comment (jekbradbury) [inline jax/interpreters/sharded_jit.py:369]:
```suggestion
  call may flow "up" the function to preceding operations as well as "down" to
```

File: jax/interpreters/sharded_jit.py
Comment (jekbradbury) [reply jax/interpreters/sharded_jit.py:393]: LGTM!

File: jax/interpreters/sharded_jit.py
Comment (skye) [reply jax/interpreters/sharded_jit.py]: Good idea, done!

**Pull request merged**
